samples/BasicEventSourcingSample/Projections/ShipInformationProjections.cs

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) IEvangelist. All rights reserved.
+// Licensed under the MIT License.
+
+using BasicEventSourcingSample.Core;
+using BasicEventSourcingSample.Projections.Models;
+using Microsoft.Azure.CosmosEventSourcing.Projections;
+using Microsoft.Azure.CosmosRepository;
+
+namespace BasicEventSourcingSample.Projections;
+
+public class ShipInformationProjections
+{
+    public class ShipCreatedHandler : IEventProjectionHandler<ShipEvents.ShipCreated>
+    {
+        private readonly IRepository<ShipInformation> _repository;
+
+        public ShipCreatedHandler(IRepository<ShipInformation> repository) =>
+            _repository = repository;
+
+        public async ValueTask HandleAsync(
+            ShipEvents.ShipCreated shipCreated,
+            CancellationToken cancellationToken = default)
+        {
+            ShipInformation info = new(
+                shipCreated.Name,
+                shipCreated.Commissioned,
+                shipCreated.OccuredUtc);
+
+            await _repository.UpdateAsync(info, cancellationToken);
+        }
+    }
+
+    public class ShipDockedHandler : IEventProjectionHandler<ShipEvents.DockedInPort>
+    {
+        private readonly IRepository<ShipInformation> _repository;
+
+        public ShipDockedHandler(IRepository<ShipInformation> repository) =>
+            _repository = repository;
+
+        public async ValueTask HandleAsync(ShipEvents.DockedInPort dockedInPort,
+            CancellationToken cancellationToken = default)
+        {
+            (string name, string port, _) = dockedInPort;
+
+            ShipInformation shipInfo = await _repository.GetAsync(
+                name,
+                nameof(ShipInformation),
+                cancellationToken);
+
+            shipInfo.LatestPort = port;
+
+            await _repository.UpdateAsync(shipInfo, cancellationToken);
+        }
+    }
+
+    public class ShipLoadedHandler : IEventProjectionHandler<ShipEvents.Loaded>
```

Comment:
In looking at these handlers, this could become pretty repetitive.  And while it does offer the most flexibility, and I we should expose this to the consumers, it would be really nice to do some of that work for the consumer. For example, imagine being able to register just the handler action:

```csharp
// They would call MapEventProjectionHandler, given the pairing of 
// ShipInformation and the Loaded event.
// They would then provide the handler delegate, which is lines 66-78.
services.MapEventProjectionHandler<ShipInformation, ShipEvents.Loaded>(
    handler: Func<IRepository<ShipInformation> repo,
                  ShipEvents.Loaded loaded,
                  CancellationToken cancellationToken,
                  ValueTask>);
```

When they map a handler in this way, we'd have to detect that and do the work of pairing the types and doing the DI for them.

```csharp
class DefaultEventProjectionHandler<TItem, TEvent> : 
    IEventProjectionHandler<TEvent> where TEvent : IPersistedEvent
    where TItem : IItem
{
    readonly IRepository<TItem> _repositoryOfT;
    readonly IEventProjectionHandlerStore _handlerStore;

    public DefaultEventProjectionHandler(
        IRepository<TItem> repositoryOfT,
        IEventProjectionHandlerStore handlerStore) => 
        (_repositoryOfT, _handlerStore) = (repositoryOfT, handlerStore);

    ValueTask IEventProjectionHandler<TEvent>.HandleAsync(
        TEvent @event,
        CancellationToken cancellationToken = default) =>
        _handlerStore.MappedFrom<TItem, TEvent>(@event, cancellationToken);
}
```

Something similar to that, this is of course pulled from my imagination so the implementation details would likely vary. The `IEventProjectionHandlerStore` would have a thread-safe, generic-type collection lookup of mapped handlers. Does that make sense?

Reply:
Once this is merged, I'd like to explore this a bit more too. I think it could be pretty slick. Note to future self, investigate this.